addons/website_slides/models/slide_channel.py

```diff
@@ -695,68 +811,123 @@ def action_redirect_to_members(self, completed=False):
+
+    def action_redirect_to_invited_members(self):
+        return self.action_redirect_to_members('invited')
+
+    def action_channel_enroll(self, channel_id=False):
+        return self.action_channel_invite(channel_id, with_enrollment=True)
+
+    def action_channel_invite(self, channel_id=False, with_enrollment=False):
+        """ Open the invitation wizard to invite and add attendees to the course.
 
+        :param channel_id: id of the channel the user wants to invite to. Used as default value.
+        :param with_enrollment: true if we want to enroll the attendees invited through the wizard.
+            False otherwise, adding attendees as 'invited', e.g. when using "Invite" action."""
+
+        template = self.env.ref(f'website_slides.mail_template_slide_channel_{"enroll" if with_enrollment else "invite"}', raise_if_not_found=False)
+        course_name = self.env['slide.channel'].sudo().browse(channel_id).name if channel_id else self.name if len(self) == 1 else ''
         local_context = dict(
             self.env.context,
-            default_channel_id=self.id,
+            default_channel_id=channel_id if channel_id else self.id if len(self) == 1 else False,
+            default_use_template=bool(template),
             default_template_id=template and template.id or False,
             default_email_layout_xmlid='website_slides.mail_notification_channel_invite',
+            default_with_enrollment=with_enrollment,
         )
+        if with_enrollment:
+            name = _('Enroll Attendees to "%(course_name)s"', course_name=course_name) if course_name else _('Enroll Attendees to a course')
+        else:
+            name = _('Invite Attendees to "%(course_name)s"', course_name=course_name) if course_name else _('Invite Attendees to a course')
+
         return {
             'type': 'ir.actions.act_window',
-            'view_mode': 'form',
+            'views': [[False, 'form']],
             'res_model': 'slide.channel.invite',
             'target': 'new',
             'context': local_context,
+            'name': name,
         }
 
     def _action_add_member(self):
-        """ Adds the logged in user in the channel members.
-        (see '_action_add_members' for more info)
+        """ Enroll the logged in user in the course. (see '_action_add_members' for more info)
+        This method is called when a user press the 'Join the course' button in front-end,
+        through '/enroll' route.
+
+        If the logged in user has been invited, then we sudo the add_members method
+        since we consider he has been granted access to the course.
+
+        Returns True if added and enrolled successfully, False otherwise."""
+        invited_res = False
+        courses_invited = self.filtered(lambda course: course.is_member_invited)
+        if courses_invited:
+            invited_res = bool(courses_invited.sudo()._action_add_members(self.env.user.partner_id))
```

Comment:
Mmmh ... that's a bit some hidden magic ... can't we sudo the call of _action_add_member in the caller code, instead of delegating it ? Will have to recheck a bit variou calls to action_add_member, for next iteration I think.

Reply:
Yep, I don't understand everything here ... but I think that maybe we should remove ``_action_add_member`` and consider ``_action_add_members`` is the only method to keep. Then callers have to adapt accordingly.

Concerning the invited_res or bool(add), have to say it is a bit difficult to follow, but the main reason seems to be the sudo . It seems simpler to me to let the controller sudo() the call to _action_add_members, especially in that case we always do it one channel at a time, meaning we know if we have to sudo or not.

Reply:
Isn't it better to perform the call as sudo in the controller once grants are verified ?

Reply:
Moved to the controller :)